doc/developer/temporal-filters.md

```diff
@@ -0,0 +1,70 @@
+# Temporal Filters
+
+Materialize is able to introduce and remove records as a function of timestamps presented as data in the records.
+This is performed using idiomatic SQL, with some restrictions, but it can be hard to understand at first.
+
+Consider the following SQL query:
+```sql
+SELECT *
+FROM data
+WHERE mz_logical_timestamp() BETWEEN data.valid_from AND data.valid_until;
+```
+The `WHERE` clause here relates two columns in the data to `mz_logical_timestamp()`, a function that returns the current logical timestamp.
+For one-shot queries `mz_logical_timestamp()` is simply evaluated and replaced by that timestamp when the query executes.
+In a *maintained* query, as for example when you call `CREATE MATERIALIZED VIEW` or `CREATE INDEX`, the value of the function changes every millisecond.
+In particular, in a maintained query any row will not be present at timestamps before its `valid_from` column nor after `valid_until` column.
+
+Materialize does not re-evaluate the predicate every millisecond for every row.
+Understanding what Materialize does do may help you understand the behavior of and restrictions on temporal filters.
+
+## Manipulating timestamped updates
+
+Materialize records *updates* to collections of data.
+Each updates has the form `(data, time, diff)`, and indicates that the occurence count of `data` changes at "logical timestamp" `time` by `diff`.
+An update occurs at its logical timestamp and is then in effect indefinitely.
+To undo an update one issues a new update with the same `data` at a future logical timestamp with a negated `diff`.
+
+We can restrict the duration of an update, from `[time, ..)` to any non-empty interval `[from, until]`, by
+1. advancing the update's timestamp to at least `from`, and
+2. introducing a new negated update at timestamp `until`.
+The first step ensures that the update does not happen until `from` and the second step ensures that the update is retracted at `until`.
+Each of the two steps is optional, if either of the `from` or `until` bounds are unspecified.
+
+This mechanism allows us to support `WHERE` clauses in which `mz_logical_timestamp()` is directly related to an expression not containing `mz_logical_timestamp()`, or conjunctions (`AND`) of these clauses.
+The following query demonstrates several example uses:
+```sql
+SELECT *
+FROM data
+WHERE mz_logical_timestamp() BETWEEN data.valid_from AND data.valid_until
```

Comment:
we should probably specify that all arguments to any inequality need to fit in unsigned 64 bits for the results to make sense / not throw range errors when you try to query the view

Reply:
Actually, idk but I did write the logic so that if the `numeric` types are outside the ranges of a `u64` it does the right thing (treats them as negative or positive infinity). Probably worth adding some tests to that effect though.

We could document this, but my read is that our choice of timestamp as a type that does not line up with `mz_logical_timestamp()` is something .. that shouldn't infect the query language at least.

Reply:
I didn't see anything that crashed but I saw some stuff that error-ed in ways that might be surprising

```
materialize=> create materialized view test2 as select * from view_one_ts where mz_logical_timestamp() < -1;
CREATE VIEW
materialize=> select * from test2;
 ts
----
(0 rows)

materialize=> select * from test2 as of -3;
ERROR:  out of range integral type conversion attempted
materialize=>
```

I think this error only makes sense if you've internalized "timestamps are positive u64s". Furthermore, anyone with negative upper bounds is just never going to see their data afaict which is fine if they know that but im not sure they do?

Reply:
I don't have super strong feelings here - just calling out that my intuition for whats going to happen with bounds that don't fit in 64 bits isnt great -- like this sequence of commands

```
materialize=> create materialized view test3 as select * from view_one_ts where mz_logical_timestamp() > 11111111111111111111111111111;
CREATE VIEW
materialize=> select * from test3;
 ts
----

  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
(11 rows)

materialize=> select * from test3 as of 1;
 ts
----
(0 rows)

materialize=> select * from test3 as of 11111111111111111111111111111;
ERROR:  out of range integral type conversion attempted
```

it seems like since this is a static view the timestamp gets forwarded to u64::max which is why the first select statement succeeds?

Reply:
I filed an issue. I *think* this is a "types" issue rather than a temporal filters issue. You would probably be just as unhappy with the fact that you can use `-1` in 
```sql
SELECT * 
FROM data
WHERE mz_logical_timestamp() = -1
```
but will get an error if you try `AS OF -1`.